File: src/00129. Sum Root to Leaf Numbers.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode *root) {
        if (NULL == root)
            return 0;
        ans = 0;
        dfs(root, 0);
        return ans;
    }
    void dfs(TreeNode *root, int sum) {
        sum *= 10;
        sum += root->val;
        if (NULL == root->left && NULL == root->right) {
            ans += sum;
        }
        if (root->left != NULL)
            dfs(root->left, sum);
        if (root->right != NULL)
            dfs(root->right, sum);
    }
private:
    int ans = 0;
};

File: src/00129. Sum Root to Leaf Numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int sum;

    public int sumNumbers(TreeNode root) {
        if (null == root) {
            return 0;
        }
        sum = 0;
        helper(0, root);
        return sum;
    }

    private void helper(int v, TreeNode root) {
        if (root.left != null) {
            helper((v + root.val) * 10, root.left);
        }
        if (root.right != null) {
            helper((v + root.val) * 10, root.right);
        }
        if (root.left == null && root.right == null) {
            sum += v + root.val;
        }
    }
}
        